hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/oauth2/WorkloadIdentityTokenProvider.java

```diff
@@ -0,0 +1,109 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.fs.azurebfs.oauth2;
+
+import java.io.File;
+import java.io.IOException;
+
+import org.apache.hadoop.thirdparty.com.google.common.base.Strings;
+import org.apache.hadoop.util.Preconditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.commons.io.FileUtils;
+
+
+/**
+ * Provides tokens based on Azure AD Workload Identity.
+ */
+public class WorkloadIdentityTokenProvider extends AccessTokenProvider {
```

Comment:
We can add tests around this class. Following is something around which we can write tests which can help prevent regressions in future:
1. how refreshing the token work.
    - We can have a protected method `getTokenTtl()` which on production code would give `ONE_HOUR`, but in test, we can mock it as per the test requirement.
    - We can mock the external call, the super call.

Reply:
Thank you for the guidance. I implemented several [unit tests](https://github.com/apache/hadoop/pull/5953/files#diff-d81fe7f19d165c02ca48b73aed52b003aeed85a60e946ac0c5ba32943ec3caff).

Reply:
I see, we have added tests around `hasEnoughTimeElapsedSinceLastRefresh`. Was expecting if token refreshing could also be tested. And, also we could see refreshing on a upper layer (still an unit test):
1. We have AccessTokenProvider object (instance of WorkloadIdentityTokenProvider ), and we could call `AccessTokenProvider` 's method `getToken()`, and then test different scenarios.

Reply:
Yes, we dont need to mock the static method. What can be done is:
we have a method: `getTokenUsingJWTAssertion()` which calls `AzureADAuthenticator
        .getTokenUsingJWTAssertion(authEndpoint, clientId, clientAssertion);`. Now, this new method is mockable, and in the test, we can give the required behavior.
Also, the methods which are mockable from the test, we add an annotation `VisibleForTesting`